driver-core/src/main/com/mongodb/connection/ByteBufferBsonOutput.java

```diff
@@ -177,7 +177,9 @@ public void truncateToPosition(final int newPosition) {
     @Override
     public void close() {
         for (final ByteBuf cur : bufferList) {
-            cur.release();
+            while(cur.getReferenceCount() > 0) {
```

Comment:
I've updated, the cause is the creation of a the `ByteBufBsonDocument` when we want to send metadata to command listeners. The issue is it is a document is backed by a `ByteBuf` and that is never released.

This is an issue with Netty as we need to release the underlying proxied netty `ByteBuf` otherwise we leak.  The initial fix was a hack to ensure the netty buffers were fully released. 

The second fix, is also slightly hacky because it takes the duplicated byteBuffers and creates the `CompositeByteBuf` and rather than duplicating the byteBuffers again (taking out another reference for netty) it gets uses the `readOnly` method.  Using `readOnly` only works because it bypasses the retaining the netty `ByteBuf`, so when `ByteBufferBsonOutput` closes it fully releases the underlying netty `ByteBuf`.

An alternative fix would be to copy the data from the ByteBuf into a RawBsonDocument but I gather that is something we want to avoid.